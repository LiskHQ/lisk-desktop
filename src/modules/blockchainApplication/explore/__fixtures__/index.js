export * from './mockBlockchainApp';
export * from './mockBlockchainAppStatistics';
