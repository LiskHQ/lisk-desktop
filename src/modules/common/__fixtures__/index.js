export * from './mockPrices';
export * from './mockCustomInfiniteQuery';
export * from './mockCcm';
export * from './mockCommandParametersSchemas';
