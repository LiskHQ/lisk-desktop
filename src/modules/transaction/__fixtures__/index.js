export * from './mockTransactions';
export * from './mockTransactionStatistics';
export * from './mockEvents';
export * from './mockFees';
