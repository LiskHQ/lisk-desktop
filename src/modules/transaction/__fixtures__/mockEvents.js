/* eslint-disable  import/prefer-default-export */
const data = new Array(30).fill({})
  .map((_, index) => ({
    id: `00${index}`,
    moduleID: `00${index + 1}`,
    moduleName: 'token',
    typeID: `00${index}`,
    data: { // Depends on event typeID
      numberOfSignatures: 2,
      mandatoryKeys: '0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a',
      optionalKeys: '0fe9a3f1a21b5530f27f87a414b549e79a940bf24fdf2b2f05e7f22aeeecc86a',
    },
    topics: [],
    block: {
      id: `625835480267616579${index}`,
      height: 8350681,
      timestamp: 28227090,
    },
  }));
export const mockEvents = {
  data,
  meta: {
    count: 15,
    offset: 0,
    total: 30,
  },
};
