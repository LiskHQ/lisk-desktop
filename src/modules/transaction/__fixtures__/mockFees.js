// eslint-disable-next-line import/prefer-default-export
export const mockFees = {
  data: {
    feeEstimatePerByte: {
      low: 0,
      medium: 0,
      high: 0,
    },
    baseFeeByID: {},
    baseFeeByName: {},
  },
  meta: {
    lastUpdate: 1660903185,
    lastBlockHeight: 412,
    lastBlockID: '65b1065e92d5d125c22f787532ff9af61a6e86eb2c5f6135d9253acbfe659879',
  },
};
