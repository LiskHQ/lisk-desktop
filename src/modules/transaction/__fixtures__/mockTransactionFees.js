export const mockTransactionFees = {
  data: {
    transactionFeeEstimates: {
      minFee: '135000',
      accountInitializationFee: {
        tokenID: '0400000000000000',
        amount: '500000',
      },
      messageFee: {
        tokenID: '0400000000000000',
        amount: '343000',
      },
    },
    dynamicFeeEstimates: {
      low: '135000',
      medium: '135000',
      high: '135000',
    },
  },
  meta: {},
};
