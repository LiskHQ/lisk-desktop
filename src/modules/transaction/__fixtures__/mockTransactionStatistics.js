/* eslint-disable import/prefer-default-export */
import moment from 'moment';

const generateTimeline = (index) => ({
  timestamp: 1556100060 + index,
  date: moment(1556100060 + index).format('YYYY-mm-DD'),
  transactionCount: `144471771933${index}`,
  volume: `144471771933${index}`,
});

const data = {
  timeline: Array(30).fill(1).map((_, idx) => generateTimeline(idx)),
  distributionByOperation: {},
  distributionByAmount: {},
};

export const mockTransactionStatistics = {
  data,
  meta: {
    count: 2,
    offset: 0,
    total: 30,
  },
};
