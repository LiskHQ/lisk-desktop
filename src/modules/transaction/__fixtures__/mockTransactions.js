const transaction = index => ({
  id: `22267562542235376${index}`,
  moduleCommandID: '2:0',
  moduleCommandName: 'token:transfer',
  nonce: '0',
  fee: '1000000',
  sender: {
    address: 'lsk24cd35u4jdq8szo3pnsqe5dsxwrnazyqqqg5eu',
    publicKey: '2ca9a7...c23079',
    name: 'genesis_51',
  },
  params: {
    // Depends on moduleCommandID
    amount: '150000000',
    recipientAddress: 'lsk24cd35u4jdq8szo3pnsqe5dsxwrnazyqqqg5eu',
    data: 'message',
  },
  block: {
    id: 35998991827805762 + index,
    height: 8350681 + index,
    timestamp: 28227090,
  },
  confirmations: 0 + index,
  executionStatus: 'pending',
  meta: {
    recipient: {
      address: 'lsk24cd35u4jdq8szo3pnsqe5dsxwrnazyqqqg5eu',
      publicKey: '2ca9a7...c23079',
      name: 'genesis_49',
    },
  },
});

const transactions = Array(30).fill(1).map((item, index) => transaction(index));

/* eslint-disable import/prefer-default-export */
export const mockTransactions = {
  meta: {
    count: 2,
    offset: 0,
    total: 30,
  },
  data: transactions,
};
