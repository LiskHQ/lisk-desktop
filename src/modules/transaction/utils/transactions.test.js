import { cryptography } from '@liskhq/lisk-client';
import { codec } from '@liskhq/lisk-codec';
import { mockNetworkStatus } from '@network/__fixtures__';
import { mockAuth } from '@auth/__fixtures__';
import { mockCommandParametersSchemas } from '../../common/__fixtures__';

import {
  Transaction,
} from './transactions';
import {
  encodeTransaction,
  fromTransactionJSON,
  toTransactionJSON,
} from './encoding';

describe.skip('Transaction', () => {
  let tokenTransfer = new Transaction();
  const recoveryPhrase = 'target cancel solution recipe vague faint bomb convince pink vendor fresh patrol';
  const { privateKey, publicKey } = cryptography.legacy.getKeys(recoveryPhrase);
  const pubkey = publicKey.toString('hex');
  const module = 'token';
  const command = 'transfer';
  const encodedTransaction = '0a05746f6b656e12087472616e73666572180020a0fb0b2a20fd061b9146691f3c56504be051175d5b76d1b1d0179c5c4370e18534c588212232470a0800000002000000011080c2d72f1a1400b1182b317a82e4b9c4a54119ced29f19b496de22207765206861766520736f6d6520696e666f726d6174696f6e2069732068657265';
  const tokenTransferParams = {
    tokenID: '0000000200000001',
    amount: '100000000', // amount in beddows
    recipientAddress: '00b1182b317a82e4b9c4a54119ced29f19b496de',
    data: 'we have some information is here',
  };
  const transactionJSON = {
    command: 'transfer',
    fee: '196000',
    id: undefined,
    module: 'token',
    nonce: '0',
    params: {
      amount: '100000000',
      data: 'we have some information is here',
      recipientAddress: '00b1182b317a82e4b9c4a54119ced29f19b496de',
      tokenID: '0000000200000001',
    },
    senderPublicKey: '0792fecbbecf6e7370f7a7b217a9d159f380d3ecd0f2760d7a55dd3e27e97184',
    signatures: [],
  };

  afterEach(() => {
    tokenTransfer = new Transaction();
  });

  describe('init', () => {
    it('should throw error when schema does not exists for module or command', () => {
      // Act
      expect(() => tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        module: 'unknown',
        command: 'unknown',
      })).toThrow('');
    });

    it('should throw error when required params not given', () => {
      // Act
      expect(() => tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
      })).toThrow('');
    });

    it('should initialise the transaction for given module and command', () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        module,
        command,
      });

      // Act
      expect(tokenTransfer.transaction).toMatchSnapshot();
    });

    it('should initialise encoded transaction', () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        encodedTransaction,
      });

      // Act
      expect(tokenTransfer.transaction).toMatchSnapshot();
    });
  });

  describe('update', () => {
    it('should be able to update command params', () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        encodedTransaction,
      });

      // Act
      const params = { ...tokenTransfer.transaction.params, data: 'updated message' };
      tokenTransfer.update({ params });
      expect(tokenTransfer.transaction.params.data).toEqual(params.data);
    });

    it('should be able to increment nonce', () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        encodedTransaction,
      });

      // Act
      const nonce = 2;
      tokenTransfer.update({ nonce });
      expect(tokenTransfer.transaction.nonce).toEqual(BigInt(nonce));
    });
  });

  describe('sign', () => {
    it('should be able to sign encoded transaction', async () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        module,
        command,
      });

      tokenTransfer.update({
        params: tokenTransferParams,
      });

      // // Act
      tokenTransfer.sign(privateKey);
      expect(tokenTransfer.transaction.signatures).toMatchSnapshot();
    });
  });

  describe('computeFee', () => {
    it('should return computed transaction fee', async () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        module,
        command,
      });

      tokenTransfer.update({
        params: tokenTransferParams,
      });

      // Act
      expect(tokenTransfer.transaction.fee).toEqual(BigInt(196000));
      tokenTransfer.update({
        params: {
          ...tokenTransferParams,
          data: 'the fee should be updated based on data field bytes length',
        },
      });
      expect(tokenTransfer.transaction.fee).toEqual(BigInt(222000));
    });
  });

  describe('encode', () => {
    it('should be able to encode transaction', async () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        module,
        command,
      });

      tokenTransfer.update({
        params: tokenTransferParams,
      });

      // Act
      expect(tokenTransfer.encode().toString('hex')).toMatchSnapshot();
    });

    it('should be able to encode transaction when params is buffer', async () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        module,
        command,
      });

      tokenTransfer.transaction.params = Buffer.alloc(0);

      // Act
      expect(encodeTransaction(tokenTransfer.transaction).toString('hex')).toEqual('0a05746f6b656e12087472616e73666572180020c8d0072a200792fecbbecf6e7370f7a7b217a9d159f380d3ecd0f2760d7a55dd3e27e971843200');
    });

    it('should allocate params empty buffer when schema is not available', async () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        module,
        command,
      });

      tokenTransfer.transaction.params = Buffer.alloc(0).toString('hex');

      // Act
      expect(encodeTransaction(tokenTransfer.transaction).toString('hex')).toEqual('0a05746f6b656e12087472616e73666572180020c8d0072a200792fecbbecf6e7370f7a7b217a9d159f380d3ecd0f2760d7a55dd3e27e971843200');
    });
  });

  describe('decode', () => {
    it('should be able decode transaction', async () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        encodedTransaction,
      });

      // Act
      expect(tokenTransfer.decode()).toMatchSnapshot();
    });
  });

  describe('toJSON', () => {
    it('should return transaction in json format', async () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        encodedTransaction,
      });
      tokenTransfer.transaction.id = Buffer.alloc(10);

      // Act
      expect(tokenTransfer.toJSON()).toEqual({
        ...transactionJSON,
        id: tokenTransfer.transaction.id.toString('hex'),
      });
    });

    it('should decode encoded params as buffer', async () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        encodedTransaction,
      });
      const encodedParams = codec.encode(
        tokenTransfer._paramsSchema,
        tokenTransfer.transaction.params,
      );
      tokenTransfer.transaction.params = encodedParams;
      tokenTransfer.transaction.id = Buffer.alloc(10).toString('hex');

      // Act
      expect(toTransactionJSON(
        tokenTransfer.transaction,
        tokenTransfer._paramsSchema,
      )).toMatchSnapshot();
    });
  });

  describe('fromJSON', () => {
    it('should return transaction object', async () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        encodedTransaction,
      });

      // Act
      expect(tokenTransfer.fromJSON()).toEqual(
        fromTransactionJSON(transactionJSON, tokenTransfer._paramsSchema),
      );
    });

    it('should decode encoded params', async () => {
      // Arrange
      tokenTransfer.init({
        pubkey,
        networkStatus: mockNetworkStatus.data,
        auth: mockAuth.data,
        commandParametersSchemas: mockCommandParametersSchemas.data,
        encodedTransaction,
      });
      const encodedParams = codec.encode(
        tokenTransfer._paramsSchema,
        tokenTransfer.transaction.params,
      );
      const trxJSON = tokenTransfer.toJSON();
      trxJSON.params = encodedParams.toString('hex');
      trxJSON.id = Buffer.alloc(10).toString('hex');

      // Act
      expect(fromTransactionJSON(trxJSON, tokenTransfer._paramsSchema)).toMatchSnapshot();
    });
  });
});
