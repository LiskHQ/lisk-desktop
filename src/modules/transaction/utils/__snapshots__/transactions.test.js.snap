// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction decode should be able decode transaction 1`] = `
Object {
  "command": "transfer",
  "fee": 196000n,
  "id": Object {
    "data": Array [
      36,
      72,
      83,
      226,
      239,
      29,
      136,
      33,
      52,
      60,
      11,
      60,
      5,
      219,
      2,
      207,
      90,
      73,
      255,
      189,
      50,
      73,
      35,
      38,
      196,
      196,
      210,
      224,
      40,
      0,
      53,
      224,
    ],
    "type": "Buffer",
  },
  "module": "token",
  "nonce": 0n,
  "params": Object {
    "amount": 100000000n,
    "data": "we have some information is here",
    "recipientAddress": Object {
      "data": Array [
        0,
        177,
        24,
        43,
        49,
        122,
        130,
        228,
        185,
        196,
        165,
        65,
        25,
        206,
        210,
        159,
        25,
        180,
        150,
        222,
      ],
      "type": "Buffer",
    },
    "tokenID": Object {
      "data": Array [
        0,
        0,
        0,
        2,
        0,
        0,
        0,
        1,
      ],
      "type": "Buffer",
    },
  },
  "senderPublicKey": Object {
    "data": Array [
      7,
      146,
      254,
      203,
      190,
      207,
      110,
      115,
      112,
      247,
      167,
      178,
      23,
      169,
      209,
      89,
      243,
      128,
      211,
      236,
      208,
      242,
      118,
      13,
      122,
      85,
      221,
      62,
      39,
      233,
      113,
      132,
    ],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`Transaction encode should be able to encode transaction 1`] = `"0a05746f6b656e12087472616e73666572180020a0fb0b2a200792fecbbecf6e7370f7a7b217a9d159f380d3ecd0f2760d7a55dd3e27e9718432470a0800000002000000011080c2d72f1a1400b1182b317a82e4b9c4a54119ced29f19b496de22207765206861766520736f6d6520696e666f726d6174696f6e2069732068657265"`;

exports[`Transaction init should initialise encoded transaction 1`] = `
Object {
  "command": "transfer",
  "fee": 196000n,
  "module": "token",
  "nonce": 0n,
  "params": Object {
    "amount": 100000000n,
    "data": "we have some information is here",
    "recipientAddress": Object {
      "data": Array [
        0,
        177,
        24,
        43,
        49,
        122,
        130,
        228,
        185,
        196,
        165,
        65,
        25,
        206,
        210,
        159,
        25,
        180,
        150,
        222,
      ],
      "type": "Buffer",
    },
    "tokenID": Object {
      "data": Array [
        0,
        0,
        0,
        2,
        0,
        0,
        0,
        1,
      ],
      "type": "Buffer",
    },
  },
  "senderPublicKey": Object {
    "data": Array [
      7,
      146,
      254,
      203,
      190,
      207,
      110,
      115,
      112,
      247,
      167,
      178,
      23,
      169,
      209,
      89,
      243,
      128,
      211,
      236,
      208,
      242,
      118,
      13,
      122,
      85,
      221,
      62,
      39,
      233,
      113,
      132,
    ],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`Transaction init should initialise the transaction for given module and command 1`] = `
Object {
  "command": "transfer",
  "fee": 125000n,
  "module": "token",
  "nonce": 0n,
  "params": Object {},
  "senderPublicKey": Object {
    "data": Array [
      7,
      146,
      254,
      203,
      190,
      207,
      110,
      115,
      112,
      247,
      167,
      178,
      23,
      169,
      209,
      89,
      243,
      128,
      211,
      236,
      208,
      242,
      118,
      13,
      122,
      85,
      221,
      62,
      39,
      233,
      113,
      132,
    ],
    "type": "Buffer",
  },
  "signatures": Array [],
}
`;

exports[`Transaction sign should be able to sign encoded transaction 1`] = `
Array [
  Object {
    "data": Array [
      158,
      18,
      126,
      106,
      165,
      90,
      134,
      169,
      36,
      129,
      221,
      40,
      16,
      101,
      232,
      157,
      195,
      87,
      140,
      148,
      117,
      113,
      59,
      29,
      79,
      100,
      148,
      88,
      84,
      10,
      239,
      144,
      206,
      1,
      57,
      157,
      132,
      64,
      48,
      160,
      66,
      160,
      197,
      12,
      172,
      251,
      92,
      64,
      218,
      202,
      180,
      123,
      180,
      125,
      210,
      78,
      89,
      178,
      189,
      219,
      87,
      25,
      58,
      5,
    ],
    "type": "Buffer",
  },
]
`;
